### PR TITLE
Adds Conditional Randomization Test (CRT) based p-value/hypothesis testing extension using TabPFN

### PR DESCRIPTION
**Pull request by msalem7777**
### Summary
This PR adds a new extension (`pval_crt`) implementing feature-level hypothesis testing using the Conditional Randomization Test (CRT) with TabPFN as a fixed predictive model.

The method tests conditional independence of individual features by resampling from an estimated conditional distribution \( X_j \mid X_{-j} \) (also modeled using TabPFN) and comparing predictive log-likelihood under the null distribution.

Both regression and classification targets are supported, and categorical vs continuous variables are handled automatically. Continuous conditional sampling is implemented via quantile predictions.

### What’s Included
- New extension under `src/tabpfn_extensions/pval_crt/`
- Example usage script under `examples/pval_crt/`
- Unit tests under `tests/` (pass with `FAST_TEST_MODE=1`)
- Python 3.9-compatible typing and explicit backend compatibility guard

### Compatibility
This extension requires the full TabPFN package and is not compatible with the tabpfn-client backend due to use of full predictive outputs and quantile predictions.

### CODE REVIEW

File: src/tabpfn_extensions/pval_crt/crt.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+        "pval_crt requires the full TabPFN package and does not support "
+        "the tabpfn-client backend."
+    ) from err
+
+from tabpfn.constants import ModelVersion
+
+from .utils import is_categorical, logp_from_full_output, logp_from_proba
+
+def tabpfn_crt(
+    X: np.ndarray,
```
Comment (LennartPurucker) [inline src/tabpfn_extensions/pval_crt/crt.py]: Could/should the test also accept DataFrames? I think in your usage above, you have the loop that maps test results to a feature, but it could be nice to pass a feature name instead of index too

Comment (LennartPurucker) [reply src/tabpfn_extensions/pval_crt/crt.py]: Maybe one step further, it might be cool to have a function that applies the test to (a subset of) all features and returns the dict per feature or a dataframe with all information for all features. 
I don't think this is a requirement for merging, but it might be cool to have!

File: src/tabpfn_extensions/pval_crt/crt.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+
+    if device is None:
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+
+    rng = np.random.RandomState(seed)
+
+    # ---------------------------
+    # Split
+    # ---------------------------
+    X_tr, X_ev, y_tr, y_ev = train_test_split(
```
Comment (LennartPurucker) [inline src/tabpfn_extensions/pval_crt/crt.py:89]: I guess for the test, we do not want to support stratification due to breaking the IID assumptions?

File: src/tabpfn_extensions/pval_crt/crt.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    )
+
+    # ---------------------------
+    # Model for y | X
+    # ---------------------------
+    y_is_cat = is_categorical(y_tr, max_unique_cat)
+    ModelY = TabPFNClassifier if y_is_cat else TabPFNRegressor
+
+    model_y = ModelY.create_default_for_version(
+        ModelVersion.V2,
```
Comment (LennartPurucker) [inline src/tabpfn_extensions/pval_crt/crt.py]: Why V2 instead of V2.5?

File: src/tabpfn_extensions/pval_crt/crt.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        device=device,
+        fit_mode="fit_with_cache",
```
Comment (LennartPurucker) [inline src/tabpfn_extensions/pval_crt/crt.py]: You might not want to use this option anymore, as it might have a bug (https://github.com/PriorLabs/TabPFN/issues/746) and might be deprecated in the future.

File: src/tabpfn_extensions/pval_crt/README.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+with foundation models using a Conditional Randomization Test (CRT).
+The implementation treats TabPFN as a fixed predictive model and uses its predictive
+log-likelihood to construct a valid test statistic.
+
+The goal is to bridge modern foundation models with classical confirmatory inference,
+particularly in settings where p-values and formal hypothesis tests are required.
+
+
+----------------------------------------------------------------------
+Motivation
```
Comment (LennartPurucker) [inline src/tabpfn_extensions/pval_crt/README.md:13]: I think it would be cool to add a "simple" explanation here for a use case. Like saying: "use this when you want to do X in practice"

**Review by LennartPurucker — COMMENTED** [review]
Heyho, cool work! 

I left some comments for you to consider below. Also, check out the comments from Gemini; some of them are quite good. 

I would be happy to approve and merge the PR once all comments are resolved.

Best,
Lennart

File: src/tabpfn_extensions/pval_crt/README.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+
+----------------------------------------------------------------------
+Empirical Validation
+----------------------------------------------------------------------
+
+We provide preliminary empirical results illustrating the calibration
+properties of the TabPFN-based Conditional Randomization Test.
+
+![CRT null p-value QQ plot](paper/qq.png)
```
[the remark this answers is not in the document]
Comment (msalem7777) [reply src/tabpfn_extensions/pval_crt/README.md]: Removed paper references from README.md.

File: src/tabpfn_extensions/pval_crt/crt.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+        if not np.all(np.isfinite(probs)):
+            i = np.argwhere(~np.isfinite(probs))[0][0]
+            print("Non-finite probs at row", i, probs[i])
+            raise ValueError("bad probs")
+        row_sums = probs.sum(axis=1)
+        if not np.allclose(row_sums, 1.0, atol=1e-6):
+            i = np.argmax(np.abs(row_sums - 1.0))
+            print("Bad prob sum at row", i, row_sums[i], probs[i])
+            raise ValueError("probabilities not normalized")
```
[the remark this answers is not in the document]
Comment (msalem7777) [reply src/tabpfn_extensions/pval_crt/crt.py]: Implemented suggested exception handling.

File: tests/test_pval_crt.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    res = tabpfn_crt(
+        X,
+        y,
+        j=2,
+        B=20,
+        test_size=0.3,
+        device="cpu",
+    )
+
+    assert "p_value" in res
```
[the remark this answers is not in the document]
Comment (msalem7777) [reply tests/test_pval_crt.py:50]: Done. Added explicit assertion verifying that irrelevant features do not trigger null rejection, while maintaining reproducibility via fixed seed.

File: src/tabpfn_extensions/pval_crt/crt.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+            f"information about the target Y beyond the remaining covariates."
+        )
+
+    summary_stmt = (
+        "\n=== Conditional Randomization Test (TabPFN) ===\n"
+        f"p-value: {p_value:.4g}\n"
+        f"{relevance_stmt}"
+    )
+
+    print(summary_stmt)
```
[the remark this answers is not in the document]
Comment (msalem7777) [reply src/tabpfn_extensions/pval_crt/crt.py]: Done. Removed print statement.

File: src/tabpfn_extensions/pval_crt/utils.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        Maximum number of unique values for categorical treatment.
+
+    Returns
+    -------
+    bool
+        True if categorical, False otherwise.
+    """
+    arr = np.asarray(arr)
+    uniq = np.unique(arr[~np.isnan(arr)])
+    return len(uniq) < max_unique
```
[the remark this answers is not in the document]
Comment (msalem7777) [reply src/tabpfn_extensions/pval_crt/utils.py]: Done.

File: tests/test_pval_crt.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        X,
+        y,
+        j=1,
+        B=20,
+        test_size=0.3,
+        device="cpu",
+    )
+
+    assert "p_value" in res
+    assert 0 <= res["p_value"] <= 1
```
[the remark this answers is not in the document]
Comment (msalem7777) [reply tests/test_pval_crt.py:28]: Done. Added explicit assertion verifying that relevant features trigger null rejection, while maintaining reproducibility via fixed seed.

File: src/tabpfn_extensions/pval_crt/crt.py (hunk 1 above)
Comment (msalem7777) [reply src/tabpfn_extensions/pval_crt/crt.py]: Done. Added dataframe support. Added feature name or index support. Added multiple feature support.

File: src/tabpfn_extensions/pval_crt/crt.py (hunk 3 above)
Comment (msalem7777) [reply src/tabpfn_extensions/pval_crt/crt.py]: Just because I was more familiar with V2 than V2.5. Added 2.5 support, and model version argument to set model version as desired.

File: src/tabpfn_extensions/pval_crt/crt.py (hunk 4 above)
Comment (msalem7777) [reply src/tabpfn_extensions/pval_crt/crt.py]: Done. Removed 'fit_mode' argument.

File: src/tabpfn_extensions/pval_crt/README.md (hunk 5 above)
Comment (msalem7777) [reply src/tabpfn_extensions/pval_crt/README.md:13]: Done. Added a "When to use" section in the README.

File: src/tabpfn_extensions/pval_crt/crt.py (hunk 2 above)
Comment (msalem7777) [reply src/tabpfn_extensions/pval_crt/crt.py:89]: Explicitly disabled stratification and left a note warning against enabling.
